### ISSUE TITLE
stopping chunkserver suspends replication

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#300]
On 2.6.0 I've changed custom goal from 

    16 16 : tst1 tst2

to

    16 16 : tst1 tst2 _

and re-loaded master. I observed how all chunkservers (except `tst1` and `tst2`) started receiving chunks until some chunkservers filled over 99%. Replication continued on chunkservers with available free space, as expected, but suddenly stopped when I've taken down server with already full chunkserver.
To summarise, replication to all chunkservers for goal `16` stopped when only one chunkserver (that was not even receiving new chunks due to lack of free space) was terminated.
Re-starting chunkserver again resumes replication to other chunkservers.

**Comment by psarna** [comment]
Hi Dmitry,

it is an issue with **REPLICATIONS_DELAY_DISCONNECT** parameter (found in *mfsmaster.cfg*, changes of its value will take place after reload, to restart required).
A quick fix is to change its value from default 3600, to, say, 30.

Explanation why it happens:
In order to prevent immediate replication after a chunkserver is disconnected for a short period of time, master waits **REPLICATIONS_DELAY_DISCONNECT** seconds before starting replication.
This time is counted per label, which means, that if chunkserver with label *label0* disconnects,
all replications related to *label0* in any way are delayed.
It is generally a very useful feature, but it has a downside - sometimes replication is delayed even if a particular server is not actively participating in it - because of per-label granularity.

I think that the wisest solution for now is to temporarily lower the value of **REPLICATIONS_DELAY_DISCONNECT** parameter just before shutting down a full chunkserver and restore it afterwards.

**Comment by onlyjob** [comment]
Hi Piotr,

Thank you for explaining. My point here is that master could be smarter and continue replication (without unnecessary suspending it) when a chunkserver is down but there are many other chunks to replicate to other chunkservers. In this particular case stopped chunkserver had no chunks to copy to other chunkservers nor it could receive any more chunks (because it was already filled). Ideally this should not hold replication so I hope that you consider such case as improvement suggestion.

We have some volatile chunkservers (e.g. on workstations that can be off overnight, etc.) so we have deliberately high `REPLICATIONS_DELAY_DISCONNECT`.

**Comment by psarna** [comment]
This improvement is on a roadmap. I hope that the `set REPLICATIONS_DELAY_DISCONNECT to low value -> disconnect desired chunkserver -> restore REPLICATIONS_DELAY_DISCONNECT value` workaround will be sufficient for you until the patch comes up.

**Comment by onlyjob** [comment]
Thanks. In the meantime it is a sufficient workaround. But why not close this issue when the patch is ready (not before)?

**Comment by psarna** [comment]
I figured out that this issue is solved, because a solution is known and planned to be done. But, since the issue belongs to you, I hereby reopen it till the patch gets merged:)

**Comment by onlyjob** [comment]
Thanks! :) I'll test and confirm once path become publicly available.
